Update snowbridge docs (#324)
snowplow/documentation: docs/destinations/forwarding-events/snowbridge/concepts/scaling/index.md

---
title: "Scaling"

> date: "2023-02-27"

Reviewer: ```suggestion
```

This does not do anything in Docusaurus

sidebar_position: 150
---

# Scaling

## Horizontal scaling

> Snowbridge is built to suit a horizontal scaling model, and you can safely deploy multiple instances of Snowbridge to consume the same input out-of-the-box (the kinesis source requires Dynamo DB tables to be created for this purpose). No addditional configuration or setup is required for the app to smoothly run across multiple instance/environment, compared to a single instance/environment.

Reviewer: ```suggestion
Snowbridge is built to suit a horizontal scaling model, and you can safely deploy multiple instances of Snowbridge to consume the same input out-of-the-box. No addditional configuration or setup is required for the app to smoothly run across multiple instance/environment, compared to a single instance/environment.

:::note

If you are using the Kinesis source, Snowbridge will automatically create Dynamo DB tables to manage multiple instances reading from the same stream.

:::
```

Author: It doesn't do that - but I've edited with a link to setup instructions.

> How to configure scaling behaviour will depend on the infrastructure you're using, and the use case you have implemented. For example, when scaling based on CPU usage, this metric will be affected by the size and shape of the data, which transformations and filters are used, and if using script transformations, the content of those scripts. 

Reviewer: ```suggestion
How to configure scaling behaviour will depend on the infrastructure you're using, and the use case you have implemented. For example, if you choose to scale based on CPU usage, note that this metric will be affected by the size and shape of the data, by the transformations and filters used, and for script transformations, by the content of the scripts.
```

> We recommend that you instrument your use case, and follow the best practices for the deployment infrastrucutre you're using to find a the scaling configuration which suits your requirements. Note that occasionally releases will improve efficiency, which in the past has had a large impact on metrics which can be used for scaling. So, it is advisable to monitor those metrics after any version update or transformation configuration change, to ensure that scaling behaves as expected.

Reviewer: ```suggestion
We recommend that you instrument your use case, and follow the best practices for the deployment infrastructure you’re using to find a the scaling configuration which suits your requirements. Note that occasionally releases will improve efficiency, which in the past has had a large impact on metrics which can be used for scaling. So, it is advisable to monitor those metrics after any version update or transformation configuration change, to ensure that scaling behaves as expected.
```

I am not sure this advice is practical enough.

> follow the best practices for the deployment infrastructure you’re using to find a the scaling configuration which suits your requirements

That just sounds like a truism, why would not someone follow the best practices? :D
Could we suggest specific ways to set up autoscaling that we recommend and/or use?

Author: I agree that it's an empty statement. However I don't have any more specific recommendation to make - there are too many variables in figuring out what scaling for your use case needs to be, it has to be a case by case decision.

I think it's hard to explain the reasoning over text - I struggled to articulate it fully in the email thread. I've removed this line for now, but I'm more than happy to hop on a zoom to understand each other on this if that's helpful. :)

> Additionally within each instance/environment, concurrency may be managed via the `concurrent_writes` setting - which provides a degree of control over throughput and resource usage. Snowbridge should consume as much data as possible, as fast as possible - a backlog of data or spike in traffic should cause the app's CPU usage to increase significantly. If spikes/backlogs do not induce this behaviour, and there are no target retries or failures (see below), then the `concurrent_writes` can be increased to induce this behaviour.

Reviewer: ```suggestion
Additionally within each instance/environment, you can manage concurrency via the `concurrent_writes` setting, which provides a degree of control over throughput and resource usage. Snowbridge should consume as much data as possible, as fast as possible — a backlog of data or spike in traffic should cause the app’s CPU usage to increase significantly. If spikes/backlogs do not induce this behaviour, and there are no target retries or failures (see below), then you can increase the `concurrent_writes`.
```

Reviewer: Shoud we move this to the Concurrency section? Seems more natural to mention after the explanation of goroutines etc.

Author: I think that too - I landed on putting this bit here because it's a piece of info that you'll specifically want if you have volume, but the app isn't processing enough data to induce the scaling you need. The thinking is that in such a scenario you might read the scaling docs to see what you're doing wrong, but not think that the concurrency section has the piece of info you actually need.

So, the way I see it is there's a bit in the scaling section saying "you can increase concurrency to eat more resources and induce scaling", then there's a section explaining what concurrency is doing.

I'm easy either way - just explaining the thought process. If you still think it's better below, I'll move it - just let me know. :)

## Concurrency

> Snowbridge is a Go application, which makes use of [goroutines](https://golangdocs.com/goroutines-in-golang). For present purposes we can think of goroutines as lightweight threads (there's dispute over how accurate this explanation is, but to explain present concepts those details aren't relevant). The source's `concurrent_writes` setting controls how many goroutines may be processing data at once, in a given instance of the app (others may exist separately, under the hood for non-data processing purposes).

Reviewer: ```suggestion
Snowbridge is a Go application, which makes use of [goroutines](https://golangdocs.com/goroutines-in-golang). You can think of goroutines as lightweight threads. The source’s `concurrent_writes` setting controls how many goroutines may be processing data at once, in a given instance of the app (others may exist separately, under the hood for non-data processing purposes).
```

> The total maximum concurrency for the entire application can be determined by multiplying `concurrent_writes` by the number of horizontal instances of the app. For example, if deployed via kubernetes pods, if I have 4 active pods with `concurrent_writes` set to 150, then at any given time there will be up to 600 concurrent goroutines that can process and send data.

Reviewer: ```suggestion
The total maximum concurrency for the entire application can be determined by multiplying `concurrent_writes` by the number of horizontal instances of the app. For example, if Snowbridge is deployed via kubernetes pods, and there are 4 active pods with `concurrent_writes` set to 150, then at any given time there will be up to 600 concurrent goroutines that can process and send data.
```

## Target scaling

> Snowbridge will attempt to send data to the target as fast as resources will allow, and so we recommend that you set up the target to scale sufficiently with the expected volume and throughput. Note that in case of failure, Snowbridge will enter an exponential retry loop, staring with a 1s delay between retries, and increasing that delay for 5 retries.

Reviewer: ```suggestion
Snowbridge will attempt to send data to the target as fast as resources will allow, so we recommend that you set up the target to scale sufficiently with the expected volume and throughput. Note that in case of failure, Snowbridge will retry sending the messages with an exponential backoff, staring with a 1s delay between retries, and increasing that delay for 5 retries.
```
Increasing how? Doubling?

Author: Good point - yes it's double. Changed. :)

> If a backlog of data builds up due to some failure - for example target downtime - then it is advisable to overprovision the target until the backlog is processed. This would only need to last until latency has fallen back to normal rates.

Reviewer: ```suggestion
If a backlog of data builds up due to some failure — for example target downtime — then we advise to overprovision the target until the backlog is processed. That’s only required until latency falls back to normal rates.
```

---
title: "Scaling"
sidebar_position: 150
---

# Scaling

## Horizontal scaling

Snowbridge is built to suit a horizontal scaling model, and you can safely deploy multiple instances of Snowbridge to consume the same input out-of-the-box. No addditional configuration or setup is required for the app to smoothly run across multiple instance/environment, compared to a single instance/environment.

> ::note

Reviewer: ```suggestion
:::note
```

> If you are using the Kinesis source, Snowbridge requires DynamoDB tables to manage consumers on the stream, as described in [the Kinesis source configuration section](/docs/destinations/forwarding-events/snowbridge/configuration/sources/kinesis.md).

Reviewer: ```suggestion
If you are using the Kinesis source, you will need to manually create a few DynamoDB tables as described in [the Kinesis source configuration section](/docs/destinations/forwarding-events/snowbridge/configuration/sources/kinesis.md). Snowbridge uses these tables to coordinate multiple instances consuming from the same stream.
```

Changing to highlight that a manual action is required. “Requires DynamoDB tables” does not convey that the user needs to intervene.

:::

How to configure scaling behaviour will depend on the infrastructure you're using, and the use case you have implemented. For example, if you choose to scale based on CPU usage, note that this metric will be affected by the size and shape of the data, by the transformations and filters used, and for script transformations, by the content of the scripts.

> Note that occasionally releases will improve efficiency, which in the past has had a large impact on metrics which can be used for scaling. So, it is advisable to monitor those metrics after any version update or transformation configuration change, to ensure that scaling behaves as expected.

Reviewer: ```suggestion
:::tip
Occasionally, new releases of Snowbridge will improve its efficiency. In the past, this has had a large impact on metrics typically used for scaling. To ensure that scaling behaves as expected, we recommend monitoring your metrics after you upgrade Snowbridge or change the transformation configuration.
:::
```

> Additionally within each instance/environment, you can manage concurrency via the `concurrent_writes` setting, which provides a degree of control over throughput and resource usage. Snowbridge should consume as much data as possible, as fast as possible — a backlog of data or spike in traffic should cause the app’s CPU usage to increase significantly. If spikes/backlogs do not induce this behaviour, and there are no target retries or failures (see below), then you can increase the `concurrent_writes`.

Reviewer: ```suggestion
In addition to configuring the number of Snowbridge instances, you can manage concurrency via the `concurrent_writes` setting (explained in the [next section](#concurrency)). This setting provides a degree of control over throughput and resource usage. Snowbridge should consume as much data as possible, as fast as possible — a backlog of data or spike in traffic should cause the app’s CPU usage to increase significantly. If spikes/backlogs do not induce this behaviour, and there are no target retries or failures (see below), then you can increase the `concurrent_writes`.
```

## Concurrency

Snowbridge is a Go application, which makes use of [goroutines](https://golangdocs.com/goroutines-in-golang). You can think of goroutines as lightweight threads. The source’s `concurrent_writes` setting controls how many goroutines may be processing data at once, in a given instance of the app (others may exist separately, under the hood for non-data processing purposes).

You can determine the total maximum concurrency for the entire application by multiplying `concurrent_writes` by the number of horizontal instances of the app. For example, if Snowbridge is deployed via kubernetes pods, and there are 4 active pods with `concurrent_writes` set to 150, then at any given time there will be up to 600 concurrent goroutines that can process and send data.

## Target scaling

> Snowbridge will attempt to send data to the target as fast as resources will allow, so we recommend that you set up the target to scale sufficiently with the expected volume and throughput. Note that in case of failure, Snowbridge will retry sending the messages with an exponential backoff, staring with a 1s delay between retries, and double that delay for 5 retries.

Reviewer: ```suggestion
Snowbridge will attempt to send data to the target as fast as resources will allow, so we recommend that you set up the target to scale sufficiently with the expected volume and throughput. Note that in case of failure, Snowbridge will retry sending the messages with an exponential backoff, staring with a 1s delay between retries, and doubling that delay for 5 retries.
```

---
title: "Scaling"
sidebar_position: 150
---

# Scaling

## Horizontal scaling

Snowbridge is built to suit a horizontal scaling model, and you can safely deploy multiple instances of Snowbridge to consume the same input out-of-the-box. No addditional configuration or setup is required for the app to smoothly run across multiple instances/environments, compared to a single instance/environment.

:::note
If you are using the Kinesis source, you will need to manually create a few DynamoDB tables as described in [the Kinesis source configuration section](/docs/destinations/forwarding-events/snowbridge/configuration/sources/kinesis.md). Snowbridge uses these tables to coordinate multiple instances consuming from the same stream.
:::

> How to configure scaling behaviour will depend on the infrastructure you're using, and the use case you have implemented. For example, if you choose to scale based on CPU usage, note that this metric will be affected by the size and shape of the data, by the transformations and filters used, and for script transformations, by the content of the scripts.

Reviewer: ```suggestion
How to configure scaling behaviour will depend on the infrastructure you’re using, and the use case you have implemented. For example, if you choose to scale based on CPU usage, note that this metric will be affected by the size and shape of the data, by the transformations and filters used, and for script transformations, by the content of the scripts.
```
